doc/BGP/BGP-supress-fib-pending.md

```diff
@@ -0,0 +1,799 @@
+
+By introducing a response channel it is required to leverage Redis Pipeline, so that the route configuration producer using Redis Pipeline with ```ProducerStateTable``` also receives route programming status responses produced by pipelined ```NotificationProducer``` which is part of ```ResponsePublisher```.
+
+On the other side, ```fpmsyncd``` does not wait for each individual route status but rather performs an asynchronous processing.
+
+The following schema represents communication methods between components:
+
+<!-- omit in toc -->
+##### Figure 6. zebra/orchagent feedback channel
+```mermaid
+%%{
+  init: {
+    "theme": "forest"
+  }
+}%%
+flowchart LR
+    zebra("</br>zebra</br></br>") -- FPM channel --> fpmsyncd("</br>fpmsyncd</br></br>")
+    fpmsyncd -- "ProducerStateTable (Pipelined)" --> RouteOrch("</br>RouteOrch</br></br>")
+    RouteOrch -. "ResponsePublisher (Pipelined)" -.-> fpmsyncd
+    fpmsyncd -. FPM channel -.-> zebra
+    RouteOrch -- "SAIRedis Bulk API" --> syncd("</br>syncd</br></br>")
+    syncd -. "SAIRedis Bulk Reply" -.-> RouteOrch
+```
+
+A ```ResponsePublisher``` must have a constructor that accepts a ```RedisPipeline``` and a flag ```buffered``` to make it use the pipelining. The constructor is similar to one in use with ```ProducerStateTable```:
+
+```c++
+ResponsePublisher::ResponsePublisher(bool buffered = false);
+```
+
+And an API to set buffered mode:
+
+```c++
+void ResponsePublisher::setBuffered(bool buffered);
+```
+
+The ```OrchDaemon``` has to flush the ```RedisPipeline``` in ```OrchDaemon::flush``` when there are no pending tasks to perform.
+
+#### 7.6. Consistency monitoring script and mitigation action
+
+Routing is a crucial part of a network switch. This feature adds additional flows in existing route processing pipeline and so along the way there might be unexpected failures leading to routes being not marked as offloaded in zebra - missed notification, race conditions leading to in-consistency, etc. It is required to monitor the consistency of routes state periodically and mitigate problems.
+
+At the moment *route_check.py* verifies routes between APPL_DB & ASIC_DB and makes sure they are in sync. In addition to that it is required to extend this script to check for consistency APPL_DB, APPL_STATE_DB and zebra by calling ```show ip route json``` and verify every route installed in ASIC_DB is marked as offloaded. The script will retry the check for 3 times every 15 sec and if zebra FIB is not in sync with ASIC_DB a mitigation action is performed. The mitigation action publishes required notifications to trigger fpmsyncd flows to send RTM_NEWROUTE message to zebra and log alerts in the syslog.

```

Comment:
Route check are currently not supported for Vrf routes. this need change in route_check script to handle VRF.
Application routes are only collected from default vrf.

Reply:
@kselvaraj2 Hi, since route_check.py does not check for VRF routes then this added check will also work only for default VRF routes. Support for VRFs needs to be considered as a separate change IMO.